### ISSUE TITLE
Fortran BP5 read finds no variables

### ISSUE THREAD

**Issue by adrianjhpc** [ornladios/ADIOS2#4232]
**Describe the bug**
The current Fortran examples in `examples/hello/bpFWriteCRead`  (`FWriter.f90` and `FReader.f90`) don't manage to write then read back the same data. Specifically, `FReader.f90` does not see any variables in the "file" that `FWriter` has produced even though a variable is evident, i.e.

```
~/bpls -d FWriter.bp/
  float    data2D  {4, 6}
    (0,0)    0 1 2 12 13 14
    (1,0)    3 4 5 15 16 17
    (2,0)    6 7 8 18 19 20
    (3,0)    9 10 11 21 22 23
```

The `FReader.f90` code was extended to see how many variables were visible in the file with this code:

```
  call adios2_available_variables(io, namestruct, ierr)
  if (ierr /= 0) stop 'adios2_available_variables returned with error'
  if (.not.namestruct%valid) stop 'adios2_available_variables returned invalid struct'
  write(*,*) 'Number of variables = ', namestruct%count
  write(*,*) 'Max name length = ', namestruct%max_name_len

  allocate(varnamelist(namestruct%count))

  call adios2_retrieve_names(namestruct, varnamelist, ierr)
  if (ierr /= 0) stop 'adios2_retrieve_names returned with error'
  if (namestruct%f2c /= 0_8) stop 'namestruct f2c pointer is not null after adios2_retrieve_names()'
  if (namestruct%valid) stop 'namestruct is not invalidated after adios2_retrieve_names()'

  do i=1,namestruct%count
     write(*,'("Var[",i2,"] = ",a12)') i, varnamelist(i)
  end do
  deallocate(varnamelist)
```

But it reports zero available as:
```
 Number of variables =            0
 Max name length =            0
```

**To Reproduce**
Compile and run the `FWriter.f90` and `FReader.f90` examples using the latest version of ADIOS main repository code.

**Expected behaviour**
Expect the `FReader` executable to find one variable, named `data2D` in the "file" that has been produced by `FWriter`.

**Desktop (please complete the following information):**
 - OS/Platform: Rocky 8.5
 - Build: Intel OneAPI 2023 compilers and MPI library

**Comment by adrianjhpc** [comment]
Note, I've tried reading the file produced by both `CppWriter` and `FWriter` in `FReader` but neither have any variables according to `FReader`. `CppReader` can read the file produced by both languages, although still has some errors when running (tries to read more data points than are in the file).

**Comment by eisenhauer** [comment]
Sorry, it looks like you've found code that has not yet been updated to BP5 semantics.  ADIOS has always had two "modes" of reading files: step-by-step using BeginStep/EndStep, and "random access" where you use SetStepSelection() to access specific input steps.  BP4 and earlier file engines blurred these two modes in that you could use the same Open statement and then do what you wanted.  The downside of this was that Open() had to read *all* the file metadata in case you wanted to use random access and not go step by step, something which was seriously problematic for really big files.  So, the BP5 engine which is the default as of ADIOS 29) introduced an additional Open mode, adios2::Mode::ReadRandomAccess (adios2_mode_read_random_access in Fortran).    If you change the Open line in FReader to specify adios2_mode_read_random_access, all should work.   Alternatively, call adios2_begin_step() before inquring of any variables.

**Comment by adrianjhpc** [comment]
Adding the step stuff fixes it. Shall I update the examples on a fork and create a PR to merge?

**Comment by eisenhauer** [comment]
> Adding the step stuff fixes it. Shall I update the examples on a fork and create a PR to merge?

That would be great.  Thank you!

**Comment by adrianjhpc** [comment]
Ok, PR created

**Comment by eisenhauer** [comment]
Closing as fixed by #4239
